File: medium/3524-find-x-value-of-array-i/solution.py
from typing import List


class Solution:
    def resultArray(self, nums: List[int], k: int) -> List[int]:
        """
        3524. Find X Value of Array I
        Time: O(n * k)
        Space: O(k)
        """
        result = [0] * k
        # cnt[x] = number of subarrays ending at the current index with
        # product % k == x
        cnt = [0] * k

        for v in nums:
            m = v % k
            nxt = [0] * k
            for x in range(k):
                if cnt[x]:
                    nxt[x * m % k] += cnt[x]
            nxt[m] += 1  # the one-element subarray [v]

            for x in range(k):
                result[x] += nxt[x]
            cnt = nxt

        return result

File: medium/3524-find-x-value-of-array-i/solution.java
/**
 * 3524. Find X Value of Array I
 * Time: O(n * k)
 * Space: O(k)
 */
class Solution {
    public long[] resultArray(int[] nums, int k) {
        long[] result = new long[k];
        // cnt[x] = number of subarrays ending at the current index with
        // product % k == x
        long[] cnt = new long[k];

        for (int v : nums) {
            int m = v % k;
            long[] next = new long[k];
            for (int x = 0; x < k; x++) {
                if (cnt[x] != 0) {
                    next[x * m % k] += cnt[x];
                }
            }
            next[m] += 1;  // the one-element subarray [v]

            for (int x = 0; x < k; x++) {
                result[x] += next[x];
            }
            cnt = next;
        }

        return result;
    }
}
